### PR TITLE
SourceManager implementation

### PR DESCRIPTION
**Pull request by mwielgus**

### CODE REVIEW

File: core/types.go [hunk 1]
```diff
@@ -45,5 +45,5 @@ type MetricSet struct {
 
 type DataBatch struct {
 	Timestamp time.Time
-	MetricSet map[string]MetricSet
+	MetricSets map[string]MetricSet
```
Comment (piosz) [inline core/types.go]: How about changing also MetricValue to plural above?

File: sources/manager.go [hunk 2]
```diff
@@ -15,21 +15,126 @@
...
 
+	"github.com/golang/glog"
 	. "k8s.io/heapster/core"
 )
 
+const (
+	MetricsScrapeTimeout = 20 * time.Second
+)
+
+//
```
Comment (piosz) [inline sources/manager.go]: I'm not sure if this is the right way to make comments;)

File: sources/manager.go [hunk 3]
```diff
@@ -15,21 +15,126 @@
...
+	return fmt.Sprintf("kubelet:%s:%d", this.host, this.port)
+}
+
+func (this *KubeletMetricsSource) ScrapeMetrics(start, end time.Time) *DataBatch {
+	var tmp DataBatch
+	return &tmp
+}
+
+//
+// Provides a list of sources to be scaped.
```
Comment (piosz) [inline sources/manager.go]: This comments is rather related to GetMetricsSources() method.

File: sources/manager.go [hunk 4]
```diff
@@ -15,21 +15,126 @@
...
+
+type KubeletProvider struct {
+}
+
+func (this *KubeletProvider) GetMetricsSources() []MetricsSource {
+	return []MetricsSource{}
+}
+
+//
+// Scrapes sources in parallel and combines the responses.
```
Comment (piosz) [inline sources/manager.go]: Same here

File: sources/manager.go [hunk 5]
```diff
@@ -15,21 +15,126 @@
...
+		if !now.Before(timeout) {
+			glog.Warningf("Failed to get all responses in time (got %d/%d)", i, len(sources))
+			break
+		}
+
+		select {
+		case dataBatch := <-responseChannel:
+			if dataBatch != nil {
+				for key, value := range dataBatch.MetricSets {
+					response.MetricSets[key] = value
```
Comment (piosz) [inline sources/manager.go]: There will be nil pointer here. Create MetricSets field in response.

Comment (mwielgus) [inline sources/manager.go]: Done

File: core/types.go (hunk 1 above)
Comment (mwielgus) [inline core/types.go]: Done

File: sources/manager.go (hunk 2 above)
Comment (mwielgus) [inline sources/manager.go]: Done.

File: sources/manager.go (hunk 3 above)
Comment (mwielgus) [inline sources/manager.go]: Done

File: sources/manager.go (hunk 4 above)
Comment (mwielgus) [inline sources/manager.go]: Done.

**Comment by piosz** [comment]
LGTM

File: sources/manager.go [hunk 6]
```diff
@@ -15,21 +15,115 @@
...
-type realSourceManager struct{}
+func (this *realSourceManager) ScrapeSources(start, end time.Time) *DataBatch {
+	sources := this.metricsSourceProvider.GetMetricsSources()
+
+	responseChannel := make(chan *DataBatch)
+	timeout := time.Now().Add(MetricsScrapeTimeout)
+
+	for _, source := range sources {
+		go func(channel chan *DataBatch, start, end, timeout time.Time) {
+			glog.Infof("Querying source: %s", source)
```
Comment (vishh) [inline sources/manager.go]: nit: I don't think we want to spam log files with info logs.

**Comment by vishh** [comment]
I don't know the overall plan behind these PRs. FYI: Certain metrics can have multiple values per container - for example, there can be multiple disks per container and we currently use resource identifiers to distinguish those disks. This can be applicable to per-cpu usage, network, etc.

**Pull request merged**